### PR TITLE
manual trigger

### PR DESCRIPTION
**Pull request by guymguym**
Signed-off-by: Guy Margalit <guymguym@gmail.com>

**Describe what this PR does**
- Add option to set `spec.trigger.manual` to some user chosen string value.
- The user can then check when the manual trigger is complete by comparing `status.lastManualSync` with the known value it has set in `spec.trigger.manual`.

**Is there anything that requires special attention?**
- No docs yet - will add once the approach is reviewed.
- No tests yet - will add once the approach is reviewed.

**Related issues:**
- None

### CODE REVIEW

**Comment by JohnStrunk** [comment]
Overall, I think this is a reasonable approach.

(I think) as written, if a manual trigger is supplied, the schedule is ignored. I'm also not sure what happens if it's in the middle of syncing and a manual sync is triggered. I suspect it may just complete the current sync and stop. We need to figure out what the desired behavior should be.

Also, if it's going to ignore the schedule, we should clear nextSyncTime since that's not accurate.

**Comment by guymguym** [comment]
@JohnStrunk I updated so that status.nextSyncTime will be unset when spec.trigger.manual is set.

**Comment by guymguym** [comment]
Regarding the question of what happens if you suddenly set `spec.trigger.manual` in the middle of a running replication. 

One case is that a user starts by setting `spec.trigger.schedule` and then decides to change to a manual trigger. In that case a running replication might start by the schedule, and the manual trigger will simply report that it completed without really starting a new replication.

Second case is that a user starts by setting `spec.trigger.manual="v1"` and while the replication is running updates the manual trigger to `"v2"`. In that case again, the manual trigger will not run a new trigger to represent the changed value.

Third case is when the trigger is empty (`nil` or `{}`) which also triggers replications constantly.

To summarize, the manual trigger can only provide a consistent knowledge on the completion of a replication when the user sets it on an idle state. The only real idle state that we have right now is if we set a manual trigger, and then complete it, and then the operator will idle. Any other case (schedule or empty trigger) might not be idle and cannot be relied on.

To me setting an empty trigger should have meant a desire to have "idle" replication. But currently this is not the case and so the only way to force idle is to set to manual and wait for it to complete and become idle, and then we can set to a new manual trigger value.

WDYT?

**Comment by guymguym** [comment]
Added a few simple test cases.

**Comment by guymguym** [comment]
I am looking for the right place in the docs to update.

The only general doc I saw on triggering/scheduling is in docs/design/custom-resources.rst:
https://github.com/backube/scribe/blob/master/docs/design/custom-resources.rst#replication-triggers

Is that a good place to add the manual trigger?

**Comment by JohnStrunk** [comment]
> I am looking for the right place in the docs to update.
> 
> The only general doc I saw on triggering/scheduling is in docs/design/custom-resources.rst:
> https://github.com/backube/scribe/blob/master/docs/design/custom-resources.rst#replication-triggers
> 
> Is that a good place to add the manual trigger?

We don't have a good place right now. How about creating a sub-page under "Usage" called <something> that has both a page for the scheduling details and for the metrics/monitoring. (i.e., move the metrics page under the new Usage>something page).
Not sure what to call the new page though: Advanced? Common options? Other? Miscellaneous? :crying_cat_face: 

Documenting the interaction between manual & schedule would be a good thing to include.

**Comment by guymguym** [comment]
@JohnStrunk Sure, let me set that doc.

**Comment by guymguym** [comment]
Anything else you think I should address in the PR?
What do you think about the behavior of setting manual I described in https://github.com/backube/scribe/pull/153#issuecomment-815556855 ?

**Comment by JohnStrunk** [comment]
> What do you think about the behavior of setting manual I described in #153 (comment) ?

It's a fine place to start. My main concern is just that it's documented. If we find that it's hard to use w/ those semantics, we can change it. I think we'll know more once we actually start to use it.

I think it's looking good. Just add the docs.

**Comment by guymguym** [comment]
@JohnStrunk I added a doc - LMK what you think. Thanks

File: docs/usage/triggers.rst [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+   # waiting for first trigger to complete...
+   while [ "$LAST_MANAUL_SYNC" != "$MANUAL" ]
+     do
+       sleep 1
+       LAST_MANAUL_SYNC=$(kubectl get replicationsource $SOURCE --template={{.status.lastManualSync}})
+       echo " - LAST_MANAUL_SYNC: $LAST_MANAUL_SYNC"
+     done
+
+   # set a second manual trigger
+   MANAUL=second
```
Comment (JohnStrunk) [inline docs/usage/triggers.rst:111]:
```suggestion
   MANUAL=second
```

File: docs/usage/triggers.rst [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+   kubectl patch replicationsources $SOURCE --type merge -p '{"spec":{"trigger":{"manual":"'$MANUAL'"}}}'
+
+   # waiting for second trigger to complete...
+   while [ "$LAST_MANAUL_SYNC" != "$MANUAL" ]
```
Comment (JohnStrunk) [inline docs/usage/triggers.rst:115]:
```suggestion
   while [ "$LAST_MANUAL_SYNC" != "$MANUAL" ]
```
... and in 117 & 118, below

**Review by JohnStrunk — COMMENTED** [review]
Just spelling, then good to go.

**Comment by guymguym** [comment]
Yikes. Fixed all typos of MANAUL=>MANUAL

**Review by JohnStrunk — APPROVED** [review]
